pkg/oc/cli/admin/release/extract.go

```diff
@@ -147,3 +207,261 @@ func (o *ExtractOptions) Run() error {
 		return opts.Run()
 	}
 }
+
+func (o *ExtractOptions) extractGit(dir string) error {
+	if err := os.MkdirAll(dir, 0750); err != nil {
+		return err
+	}
+
+	release, err := NewInfoOptions(o.IOStreams).LoadReleaseInfo(o.From)
+	if err != nil {
+		return err
+	}
+
+	cloner := &git{}
+
+	hadErrors := false
+	alreadyExtracted := make(map[string]string)
+	for _, ref := range release.References.Spec.Tags {
+		repo := ref.Annotations["io.openshift.build.source-location"]
+		commit := ref.Annotations["io.openshift.build.commit.id"]
+		if len(repo) == 0 || len(commit) == 0 {
+			if glog.V(2) {
+				glog.Infof("Tag %s has no source info", ref.Name)
+			} else {
+				fmt.Fprintf(o.ErrOut, "warning: Tag %s has no source info\n", ref.Name)
+			}
+			continue
+		}
+		if oldCommit, ok := alreadyExtracted[repo]; ok {
+			if oldCommit != commit {
+				fmt.Fprintf(o.ErrOut, "warning: Repo %s referenced more than once with different commits, only checking out the first reference\n", repo)
+			}
+			continue
+		}
+		alreadyExtracted[repo] = commit
+
+		basePath, err := sourceLocationAsRelativePath(dir, repo)
+		if err != nil {
+			return err
+		}
+
+		var git *git
+		fi, err := os.Stat(basePath)
+		if err != nil {
+			if !os.IsNotExist(err) {
+				return err
+			}
+			if err := os.MkdirAll(basePath, 0750); err != nil {
+				return err
+			}
+		} else {
+			if !fi.IsDir() {
+				return fmt.Errorf("repo path %s is not a directory", basePath)
+			}
+		}
+		git, err = cloner.ChangeContext(basePath)
+		if err != nil {
+			if err != noSuchRepo {
+				return err
+			}
+			glog.V(2).Infof("Cloning %s ...", repo)
+			if err := git.Clone(repo, o.Out, o.ErrOut); err != nil {
+				hadErrors = true
+				fmt.Fprintf(o.ErrOut, "error: cloning %s: %v\n", repo, err)
+				continue
+			}
+		}
+		glog.V(2).Infof("Checkout %s from %s ...", commit, repo)
+		if err := git.CheckoutCommit(repo, commit); err != nil {
+			hadErrors = true
+			fmt.Fprintf(o.ErrOut, "error: checking out commit for %s: %v\n", repo, err)
+			continue
+		}
+	}
+	if hadErrors {
+		return kcmdutil.ErrExit
+	}
+	return nil
+}
+
+type git struct {
```

Comment:
Use `github.com/openshift/library-go/pkg/git` instead

Reply:
I don't like that code, and it's not actually making it easier to find bugs.

Reply:
Also, I need access to the underlying shim, which that package doesn't offer, and I don't want to refactor that now.

Reply:
I will however split it out and leave a TODO to look for overlap.